xayn-ai/src/ranker/context.rs

```diff
@@ -19,13 +21,82 @@ impl Context {
         Self { pos_avg, neg_max }
     }
 
-    /// Calculates context value from given LTR score, positive distance, negative distance and similarity.
-    pub(crate) fn calculate(&self, pos: f32, neg: f32) -> f32 {
+    /// Calculates context value from given positive distance, negative distance.
+    pub(crate) fn calculate_score(&self, pos: f32, neg: f32) -> f32 {
         let frac_pos = (self.pos_avg > 0.)
-            .then(|| (5. + pos / self.pos_avg).recip())
-            .unwrap_or(5.);
-        let frac_neg = (5. + (self.neg_max - neg)).recip();
+            .then(|| (1. + pos / self.pos_avg).recip())
+            .unwrap_or(1.);
+        let frac_neg = (1. + (self.neg_max - neg)).recip();
 
-        (frac_pos + frac_neg) / 10.
+        (frac_pos + frac_neg) / 2.
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        data::document_data::CoiComponent,
+        ranker::{context::Context, util::Id},
+        CoiId,
+    };
+    use test_utils::assert_approx_eq;
+
+    #[allow(clippy::eq_op)]
+    #[test]
+    fn test_calculate() {
+        let calc = Context {
+            pos_avg: 4.,
+            neg_max: 8.,
+        };
+
+        // In the `assert_approx_eq!`s below, the result expectations are
+        // expressions instead of constants to make it easier to understand how
+        // the values come into existence. The format is (frac_pos / frac_neg) / 2.
+
+        let cxt = calc.calculate_score(0., calc.neg_max);
+        assert_approx_eq!(f32, cxt, (1. + 1.) / 2.);
+
+        let cxt = calc.calculate_score(1., calc.neg_max);
+        assert_approx_eq!(f32, cxt, ((4. / 5.) + 1.) / 2.);
+
+        let cxt = calc.calculate_score(calc.pos_avg, calc.neg_max);
+        assert_approx_eq!(f32, cxt, (1. / 2. + 1.) / 2.);
+
+        let cxt = calc.calculate_score(8., 7.);
+        assert_approx_eq!(f32, cxt, ((1. / 3.) + (1. / 2.)) / 2.);
+    }
+
+    #[test]
+    fn test_from_cois() {
+        let cois = vec![
+            (
+                Id::from_rand(),
```

Comment:
we can also use `Id::from_u128(0)` because for the context calculation it doesn't matter if the id is unique